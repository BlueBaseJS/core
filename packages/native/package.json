{
  "name": "bluerain-boilerplate-native",
  "version": "0.3.0",
  "description": "Hello BlueRain!",
  "author": "BlueEast Team <code@blueeast.com>",
  "private": true,
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "start": "exp start"
  },
  "dependencies": {
    "@blueeast/bluerain-os": "git://github.com/BlueEastCode/bluerain-os.git#feature/custom_platform",
    "expo": "19.0.1",
    "react": "16.0.0-alpha.12",
    "react-dom": "^16.0.0-alpha.12",
    "react-native": "0.46.4",
    "react-native-windows": "^0.33.0",
    "reactxp": "^0.42.0-rc.25",
    "reactxp-navigation": "1.0.5",
    "reactxp-video": "^0.2.0"
  },
  "devDependencies": {
    "@types/node": "^7.0.12",
    "@types/react": "15.0.39",
    "@types/react-native": "0.46.7",
    "@types/webpack": "^2.2.14",
    "source-map-loader": "^0.1.6"
  }
}
