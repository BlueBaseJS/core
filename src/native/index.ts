export * from './ActivityIndicator';
export * from './Button';
export * from './Image';
export * from './Text';
export * from './View';
