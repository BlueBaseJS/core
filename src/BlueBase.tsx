import { Analytics, Logger } from './api';
import {
	ComponentCollection,
	ComponentRegistry,
	ConfigCollection,
	ConfigRegistry,
	HookNestedCollection,
	HookRegistry,
	PluginCollection,
	PluginRegistry,
	ThemeCollection,
	ThemeRegistry,
} from './registries';
import { BlueBaseProvider } from './Context';
import React from 'react';
import { ThemeProvider } from './themes';
import { getComponent } from './getComponent';
import systemHooks from './hooks';

export interface BootOptions {
	components: ComponentCollection,
	configs: ConfigCollection,
	hooks: HookNestedCollection,
	plugins: PluginCollection,
	themes: ThemeCollection,
	// routes: Plugin[]
}

export class BlueBase {

	// APIs
	public Analytics = new Analytics(this);
	public Logger = new Logger(this);

	// Registries
	public Components = new ComponentRegistry(this);
	public Configs = new ConfigRegistry(this);
	public Hooks = new HookRegistry(this);
	public Plugins = new PluginRegistry(this);
	public Themes = new ThemeRegistry(this);

	// Flags
	public booted = false;

	private bootOptions: BootOptions = {
		components: {},
		configs: {},
		hooks: {},
		plugins: [],
		themes: [],
	};

	public async boot(options?: Partial<BootOptions> & { children?: React.ReactNode }) {

		this.bootOptions = { ...this.bootOptions, ...options };

		await this.Hooks.registerNestedCollection(systemHooks);
		await this.Hooks.registerNestedCollection(this.bootOptions.hooks);

		await this.Hooks.run('bluebase.boot', this.bootOptions);

		const SystemApp = getComponent('SystemApp');

		// Set View
		// const SystemApp = this.Components.resolve('SystemApp');
		// SystemApp = await this.Hooks.run('bluebase.system.app', SystemApp);

		const BlueBaseRoot = () => (
			<BlueBaseProvider value={this}>
				<ThemeProvider>
					<SystemApp>
						{options && options.children}
					</SystemApp>
				</ThemeProvider>
			</BlueBaseProvider>
		);

		this.booted = true;

		return BlueBaseRoot;
	}
}