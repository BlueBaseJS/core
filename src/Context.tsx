import { BlueBase } from './BlueBase';
import { createContext } from 'react';

export const BlueBaseContext: React.Context<BlueBase> = createContext(undefined as any);
export const BlueBaseProvider = BlueBaseContext.Provider;
export const BlueBaseConsumer = BlueBaseContext.Consumer;
