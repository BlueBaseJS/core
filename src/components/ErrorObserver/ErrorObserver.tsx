import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import { MaybeRenderPropChildren } from '../../utils';
import React from 'react';

const MISSING_ERROR = 'An unknown error occured.';

export interface ErrorObserverProps {
	error?: Error,
	checkError?: (props: ErrorObserverProps & any) => Error,
	errorComponent?: React.ComponentType<any>;
	children?: MaybeRenderPropChildren;
}

export interface ErrorObserverState {
	readonly error?: Error,
}

export class ErrorObserver extends React.PureComponent<ErrorObserverProps, ErrorObserverState> {

	static contextType = BlueBaseContext;

	public static defaultProps: Partial<ErrorObserverProps> = {
		checkError: (props) => props.error,
	};

	readonly state: ErrorObserverState = {
		error: (this.props.checkError) ? this.props.checkError(this.props) : undefined,
	};

	componentDidCatch(error: Error | null) {
		this.setState({
			error: error || new Error(MISSING_ERROR)
		});
	}

	render() {

		// FIXME: remove typecasting, added because current react typings don't seem to support this.context
		const BB: BlueBase = (this as any).context;

		const { error } = this.state;
		const { children } = this.props;

		if (error) {
			BB.Logger.error(error);
			const Error = this.props.errorComponent || BB.Components.ErrorState;
			return React.createElement(Error, { error });
		}

		// 'children' as a function, 'render prop' pattern
		if (typeof children === 'function') {
			return (children as any)();
		}

		return children;
	}
}
