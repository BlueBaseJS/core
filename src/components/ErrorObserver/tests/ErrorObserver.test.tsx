// import * as Component from '../..';
import * as Native from '../../../native';
import { BlueBase } from '../../../BlueBase';
import { BlueBaseProvider } from '../../../Context';
import { ErrorObserver } from '../ErrorObserver';
import React from 'react';
import TestRenderer from 'react-test-renderer';

beforeEach(() => {
	jest.resetModules();
});

const BB = new BlueBase();

describe('ErrorObserver', () => {
	const ErrorObserverWithProvider = (props: any) => (
		<BlueBaseProvider value={BB}>
			<ErrorObserver {...props}/>
		</BlueBaseProvider>
	);

	test(`Snapshot ErrorObserver`, () => {
		const component = TestRenderer.create(
			<ErrorObserverWithProvider>
				<Native.Text>Hello</Native.Text>
			</ErrorObserverWithProvider>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot ErrorObserver with error`, () => {
		const component = TestRenderer.create(
			<ErrorObserverWithProvider error={{ name: '404', message: 'no page found' }}
				errorComponent={() => <Native.Text>Error</Native.Text>}>
				<Native.Text>Hello</Native.Text>
			</ErrorObserverWithProvider>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot ErrorObserver with child as function`, () => {
		const component = TestRenderer.create(
			<ErrorObserverWithProvider>
				{
					() => <Native.Text>Hello</Native.Text>
				}
			</ErrorObserverWithProvider>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

});
