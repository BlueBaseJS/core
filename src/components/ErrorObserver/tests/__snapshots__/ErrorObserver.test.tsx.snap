// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorObserver Snapshot ErrorObserver 1`] = `
<Text>
  Hello
</Text>
`;

exports[`ErrorObserver Snapshot ErrorObserver with child as function 1`] = `
<Text>
  Hello
</Text>
`;

exports[`ErrorObserver Snapshot ErrorObserver with error 1`] = `
<Text>
  Error
</Text>
`;
